flocker/node/agents/blockdevice.py

```diff
@@ -112,6 +127,201 @@ class UnattachedVolume(VolumeException):
+)
+
+UNMOUNT_BLOCK_DEVICE = ActionType(
+    u"agent:blockdevice:unmount",
+    [VOLUME],
+    [],
+    u"A block-device-backed dataset is being unmounted.",
+)
+
+DETACH_VOLUME = ActionType(
+    u"agent:blockdevice:detach_volume",
+    [VOLUME],
+    [],
+    u"The volume for a block-device-backed dataset is being detached."
+)
+
+DESTROY_VOLUME = ActionType(
+    u"agent:blockdevice:destroy_volume",
+    [VOLUME],
+    [],
+    u"The volume for a block-device-backed dataset is being destroyed."
+)
+
+
+def _logged_statechange(cls):
+    """
+    Decorate an ``IStateChange.run`` implementation with partially automatic
+    logging.
+
+    :param cls: An ``IStateChange`` implementation which also has an
+        ``_action`` attribute giving an Eliot action that should be used to log
+        its ``run`` method.
+
+    :return: ``cls``, mutated so that its ``run`` method is automatically run
+        in the context of its ``_action``.
+    """
+    original_run = cls.run
+    # Work-around https://twistedmatrix.com/trac/ticket/7832
+    try:
+        original_run.__name__ = original_run.methodName
+    except AttributeError:
+        pass
+
+    @wraps(original_run)
+    def run(self, deployer):
+        with self._action:
+            # IStateChange.run nominally returns a Deferred.  Hook it up to the
+            # action properly.  Do this as part of FLOC-1549 or maybe earlier.
+            return original_run(self, deployer)
+
+    cls.run = run
+    return cls
+
+
+class BlockDeviceVolume(PRecord):
+    """
+    A block device that may be attached to a host.
+
+    :ivar unicode blockdevice_id: The unique identifier of the block device.
+    :ivar int size: The size, in bytes, of the block device.
+    :ivar unicode host: The IP address of the host to which the block device is
+        attached or ``None`` if it is currently unattached.
+    :ivar UUID dataset_id: The Flocker dataset ID associated with this volume.
+    """
+    blockdevice_id = field(type=unicode, mandatory=True)
+    size = field(type=int, mandatory=True)
+    host = field(type=(unicode, type(None)), initial=None)
+    dataset_id = field(type=UUID, mandatory=True)
+
+
+@_logged_statechange
+@with_cmp(["change"])
+class DestroyBlockDeviceDataset(proxyForInterface(IStateChange, "change")):
+    """
+    Destroy the volume for a dataset with a primary manifestation on the node
+    where this state change runs.
+
+    :ivar IStateChange change: The real implementation of this state change.
+    :ivar volume: See ``__init__``.
+    """
+    def __init__(self, volume):
+        """
+        :param BlockDeviceVolume volume: The volume which will be destroyed.
+        """
+        self.volume = volume
+        sequence = Sequentially(changes=[
+            UnmountBlockDevice(volume=volume),
+            DetachVolume(volume=volume),
+            DestroyVolume(volume=volume),
+        ])
+
+        super(DestroyBlockDeviceDataset, self).__init__(sequence)
+
+    @property
```

Comment:
And I'd rather a quick version of FLOC-1591 were done than this. If you *do* stick to this `_action` isn't a great name, it's super confusing. `_eliot_action`?